crates/js_api/src/gui/order_operations.rs

```diff
@@ -0,0 +1,293 @@
+use super::*;
+use alloy::{
+    primitives::{utils::parse_units, Bytes, U256},
+    sol_types::SolCall,
+};
+use alloy_ethers_typecast::{
+    multicall::IMulticall3::{aggregate3Call, Call3},
+    transaction::{ReadContractParameters, ReadableClientHttp},
+};
+use rain_orderbook_app_settings::{order::OrderIO, orderbook::Orderbook};
+use rain_orderbook_bindings::IERC20::decimalsCall;
+use rain_orderbook_common::{
+    add_order::AddOrderArgs, deposit::DepositArgs, transaction::TransactionArgs,
+};
+use std::{collections::HashMap, str::FromStr, sync::Arc};
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Tsify)]
+#[tsify(into_wasm_abi, from_wasm_abi)]
+pub struct TokenAllowance {
+    #[tsify(type = "string")]
+    token: Address,
+    #[tsify(type = "string")]
+    allowance: U256,
+}
+impl_wasm_traits!(TokenAllowance);
+
+#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Tsify)]
+#[tsify(into_wasm_abi, from_wasm_abi)]
+pub struct ApprovalCalldata {
+    #[tsify(type = "string")]
+    token: Address,
+    #[tsify(type = "string")]
+    calldata: Bytes,
+}
+impl_wasm_traits!(ApprovalCalldata);
+
+#[wasm_bindgen]
+impl DotrainOrderGui {
+    fn get_orderbook(&self) -> Result<Arc<Orderbook>, GuiError> {
+        self.deployment
+            .deployment
+            .as_ref()
+            .order
+            .as_ref()
+            .orderbook
+            .as_ref()
+            .ok_or(GuiError::OrderbookNotFound)
+            .cloned()
+    }
+
+    fn get_all_vaults(&self) -> Vec<OrderIO> {
+        let mut all_vaults: Vec<_> = self.deployment.deployment.order.inputs.clone();
+        let filtered_outputs: Vec<OrderIO> = self
+            .deployment
+            .deployment
+            .order
+            .outputs
+            .iter()
+            .filter(|output| !all_vaults.contains(*output))
+            .cloned()
+            .collect();
+        all_vaults.extend(filtered_outputs);
+        all_vaults
+    }
+
+    fn get_deposits_as_map(&self) -> HashMap<Address, String> {
+        self.get_deposits()
+            .into_iter()
+            .map(|d| (d.address, d.amount))
+            .collect()
+    }
+
+    async fn get_token_decimals(&self, rpc_url: &str, token: Address) -> Result<u8, GuiError> {
+        let client = ReadableClientHttp::new_from_url(rpc_url.to_string())?;
+        let decimals = client
+            .read(ReadContractParameters {
+                address: token,
+                call: decimalsCall {},
+                block_number: None,
+            })
+            .await?
+            ._0;
+        Ok(decimals)
+    }
+
+    async fn get_deposit_args(
+        &self,
+        rpc_url: &str,
+        deposits_map: &HashMap<Address, String>,
+        order_io: &OrderIO,
+    ) -> Result<DepositArgs, GuiError> {
+        let decimals = if let Some(decimals) = order_io.token.decimals {
```

Comment:
it's not really safe to use the decimals in the yaml - i think we should always check onchain

what would be good is to store all the token info (name/symbol/decimals) in memory, and allow it to be prefetched. then here we would either use the info already in memory, or fetch if we have to